### ISSUE TITLE
Feature request: Spy 

### ISSUE THREAD

**Issue by marceloboeira** [waterlink/mocks.cr#5]
As discussed [here](https://groups.google.com/forum/#!topic/crystal-lang/WNuwiRXJ8E0), I think it would be nice to have the possibility to spy on objects. 
- [x] `expect(object).to receive(:method).and_return(10)`

or even
- [x] `expect(object).to have_received(:method).and_return(10)`

It would be very useful!

**Comment by waterlink** [comment]
So, `.to receive(...)` will setup an expectation that should be verified at the end of each example. While `.to have_received(...)` will verify right away. It means, that implementation of `have_received` will need to simply implement new matcher for `spec2` (and/or just stdlib's `spec`) and after requiring it, everything should just work. That makes it much simpler to implement, that is why it is worth to implement it as the first thing. Later `.to receive(...)` can be implemented, after relevant changes to `spec2` to allow setting up of an expectations, that will be verified after the example.

Additionally, now that I think about it:
- `mocks` library should implement simple matcher class, with simple understandable interface and no magic and macro going.
- something like `mocks-spec` and `mocks-spec2` should implement macro/method versions of this matcher and plug them into relevant libraries.
- `spec2` needs to add a mechanism to run and register late expectations.

WDYT about such plan?

**Comment by marceloboeira** [comment]
@waterlink I thought I had opened this issue on the spec2 repository, sorry. haha (PS: you own too many crystal libs)

This is a good plan, but if you don't actually need the method to run, yet, just confirm it has been called, still makes sense to be on the "mocks" lib?

**Comment by waterlink** [comment]
So, I have:
- Implemented `HaveReceivedExpectation` together with macro `have_received`, that is compatible with stdlib `spec`'s expectation protocol. Usage example can be seen here: https://github.com/waterlink/mocks.cr#havereceived-expectation-usage-with-stdlib-spec
- Created new library [`spec2-mocks`](https://github.com/waterlink/spec2-mocks.cr), which creates adapter for `HaveReceivedExpectation`, so that it works with `spec2`'s matcher protocol. Usage example can be found here: https://github.com/waterlink/spec2-mocks.cr#usage

**Comment by waterlink** [comment]
`.to receive(...)` I will do a bit later, since it will require some re-design on `spec2` side.

**Comment by marceloboeira** [comment]
@waterlink Great! Thanks a lot!

**Comment by marceloboeira** [comment]
@waterlink currently I'm facing some problems when using HaveReceivedExpectation, it works well for objects, but not quite for Module methods.

For instance, this:

``` crystal
create_module_mock Notifier do
  mock self.notify(title)
end

describe "A" do
  it "expects to have received" do
     Notifier.notify("hello")

     expect(Notifier).to have_received(self.notify("hello"))
  end
end
```

generates the following error:

```
Error in macro 'it' /.../spec2-nc/libs/spec2/context.cr:10, line 2:

  1.       instance.examples << ::Spec2::Example.new(self, "notifies with error message", "expanded macro: describe", 38) do |ctx|
  2.         (ctx as self).run_hook do
  3.   reporter.report
  4.   (expect(Notifier)).to(have_received(self.notify("hello")))
  5. end
  6.       end
  7.

        (ctx as self).run_hook do
                      ^~~~~~~~

instantiating 'Spec2__Spec2__reporters__nc_1::Spec2__With_failed_examples_2#run_hook()'
in macro 'it' /Users/marceloboeira/Development/marceloboeira/spec2-nc/libs/spec2/context.cr:10, line 2:

  1.       instance.examples << ::Spec2::Example.new(self, "notifies with error message", "expanded macro: describe", 38) do |ctx|
  2.         (ctx as self).run_hook do
  3.   reporter.report
  4.   (expect(Notifier)).to(have_received(self.notify("hello")))
  5. end
  6.       end
  7.

        (ctx as self).run_hook do
                      ^~~~~~~~

instantiating 'Spec2__Spec2__reporters__nc_1::Spec2__With_failed_examples_2#run_hook()'
in macro 'it' /../spec2-nc/libs/spec2/context.cr:10, line 4:

  1.       instance.examples << ::Spec2::Example.new(self, "notifies with error message", "expanded macro: describe", 38) do |ctx|
  2.         (ctx as self).run_hook do
  3.   reporter.report
  4.   (expect(Notifier)).to(have_received(self.notify("hello")))
  5. end
  6.       end
  7.

  (expect(Notifier)).to(have_received(self.notify("hello")))
                             ^~

instantiating 'Spec2::Expectation(Notifier:Module)#to(Spec2::Mocks::HaveReceived)'
in ./libs/spec2/expectation.cr:12: instantiating 'Spec2::Mocks::HaveReceived#match(Notifier:Module)'

      return if matcher.match(actual)
                        ^~~~~

in ./libs/spec2-mocks/spec2-mocks.cr:14: instantiating 'Mocks::HaveReceivedExpectation#match(Notifier:Module)'

      unwrap.match(value)
             ^~~~~

in ./libs/mocks/mocks/have_received_expectation.cr:7: instantiating 'oid()'

      method.received?(oid, @receive.args)
                       ^~~

in ./libs/mocks/mocks/have_received_expectation.cr:25: instantiating 'Mocks::Registry::ObjectId:Class#build((Nil | Notifier:Module))'

      Registry::ObjectId.build(@target)
                         ^~~~~

in ./libs/mocks/mocks/registry.cr:27: undefined method 'object_id' for Notifier:Module (compile-time type is (Nil | Notifier:Module))

        new(object.object_id)
                   ^~~~~~~~~

================================================================================

Notifier:Module trace:

  ./libs/mocks/mocks/registry.cr:26

          def self.build(object)
```

I believe there is something wrong when registering the module, it was expecting an object, or something that could respond to `object.object_id`, yet I'm not sure.

**Comment by marceloboeira** [comment]
@waterlink thanks!

**Comment by waterlink** [comment]
@marceloboeira I have released 0.4.1 with the fix

**Comment by waterlink** [comment]
I have released `0.3.0` version of `spec2-mocks` that has `expect(...).to receive(...)` functionality in.
